internal/resources/iampolicy/resource_iam_policy.go

```diff
@@ -0,0 +1,505 @@
+	}
+
+	return policyList, err
+}
+
+// resourceIAMPolicyRead returns the intersection between binding list in terraform state and TMC server.
+func resourceIAMPolicyRead(_ context.Context, d *schema.ResourceData, m interface{}) (diags diag.Diagnostics) {
+	config, _ := m.(authctx.TanzuContext)
+
+	var (
+		meta         *objectmetamodel.VmwareTanzuCoreV1alpha1ObjectMeta
+		rbServerList []*iammodel.VmwareTanzuCoreV1alpha1PolicyRoleBinding
+		calRBList    []*iammodel.VmwareTanzuCoreV1alpha1PolicyRoleBinding
+	)
+
+	scopedFullnameData := constructScope(d)
+	rbStateList := constructRoleBindingList(d)
+
+	if scopedFullnameData == nil {
+		return diag.Errorf("unable to get Role Bindings; Scope full name: %v is empty", d)
+	}
+
+	policyList, err := retrieveRoleBindingListFromServer(config, scopedFullnameData)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	// when iam policy resource is empty: no role bindings are existing, this is equivalent to not found condition.
+	if len(policyList) == 0 {
+		d.SetId("")
+		return
+	}
+
+	d.SetId(d.State().ID)
+
+	if err := d.Set(common.MetaKey, common.FlattenMeta(meta)); err != nil {
+		return diag.FromErr(err)
+	}
+
+	// iterate over the policy lists: state and server, and store the intersection of the two
+	for _, policy := range policyList {
+		if policy.Meta.UID == d.State().ID {
+			rbServerList = append(rbServerList, policy.RoleBindings...)
+		}
+	}
+
+	for _, stateRB := range rbStateList {
+		for _, serverRB := range rbServerList {
+			if stateRB.Role == serverRB.Role {
+				calRBList = append(
+					calRBList,
+					&iammodel.VmwareTanzuCoreV1alpha1PolicyRoleBinding{
+						Role:     serverRB.Role,
+						Subjects: getIntersectionOfSubs(stateRB.Subjects, serverRB.Subjects),
+					},
+				)
+			}
+		}
+	}
+
+	if err := d.Set(roleBindingsKey, flattenRoleBindingList(calRBList)); err != nil {
+		return diag.FromErr(err)
+	}
+
+	return diags
+}
+
+func getIntersectionOfSubs(
+	state, server []*iammodel.VmwareTanzuCoreV1alpha1PolicySubject,
+) []*iammodel.VmwareTanzuCoreV1alpha1PolicySubject {
+	var newList []*iammodel.VmwareTanzuCoreV1alpha1PolicySubject
+
+	subjectIntersect := make(map[string]*iammodel.VmwareTanzuCoreV1alpha1PolicySubject)
+	for _, each := range state {
+		subjectIntersect[each.Name+"_"+string(*each.Kind)] = each
+	}
+
+	for _, sub := range server {
+		if each, ok := subjectIntersect[sub.Name+"_"+string(*sub.Kind)]; ok {
+			newList = append(newList, each)
+		}
+	}
+
+	return newList
+}
+
+func getBindingDeltaListForUpdate(oldRBL, newRBL, serverRBL []*iammodel.VmwareTanzuCoreV1alpha1PolicyRoleBinding) (blData []*iammodel.VmwareTanzuCoreV1alpha1PolicyBindingDelta) {
+	subjectIntersect := make(map[string]iammodel.VmwareTanzuCoreV1alpha1PolicyBindingDeltaOpType)
+
+	for _, old := range oldRBL {
```

Comment:
can you merge all the three loops into a single function which is called three times ? to reduce the code duplication